src/encoder/stateless/vp9/predictor.rs

```diff
@@ -0,0 +1,169 @@
+use crate::encoder::stateless::Predictor;
+use crate::encoder::FrameMetadata;
+
+#[derive(Clone)]
+pub enum PredictionStructure {
+    /// Simplest prediction structure, suitable eg. for RTC. Interframe is produced at the start of
+    /// the stream and every time when [`limit`] frames are reached. Following interframe frames
+    /// are frames relying solely on the last frame.
+    LowDelay { limit: u16 },
+}
+
+/// See [`PredictionStructure::LowDelay`]
+pub(super) struct LowDelay<P, R> {
+    queue: VecDeque<(P, FrameMetadata)>,
+
+    references: VecDeque<Rc<R>>,
+
+    counter: usize,
+
+    /// Encoder config
+    config: EncoderConfig,
+}
+
+impl<P, R> LowDelay<P, R> {
+    pub(super) fn new(config: EncoderConfig) -> Self {
+        Self {
+            queue: Default::default(),
+            references: Default::default(),
+            counter: 0,
+            config,
+        }
+    }
+
+    fn request_keyframe(
+        &mut self,
+        input: P,
+        input_meta: FrameMetadata,
+    ) -> EncodeResult<Vec<BackendRequest<P, R>>> {
+        log::trace!("Requested keyframe timestamp={}", input_meta.timestamp);
+
+        let header = Header {
+            frame_type: FrameType::KeyFrame,
+            show_frame: true,
+            error_resilient_mode: true,
+            width: self.config.resolution.width,
+            height: self.config.resolution.height,
+            render_and_frame_size_different: false,
+            render_width: self.config.resolution.width,
+            render_height: self.config.resolution.height,
+            intra_only: true,
+            refresh_frame_flags: 0x01,
+            ref_frame_idx: [0, 0, 0],
+
+            ..Default::default()
+        };
+
+        let request = BackendRequest {
+            header,
+            input,
+            input_meta,
+            last_frame_ref: None,
+            golden_frame_ref: None,
+            altref_frame_ref: None,
+            bitrate: self.config.bitrate.clone(),
+            coded_output: Vec::new(),
+        };
+
+        self.counter += 1;
+
+        Ok(vec![request])
+    }
+
+    fn request_interframe(
+        &mut self,
+        input: P,
+        input_meta: FrameMetadata,
+    ) -> EncodeResult<Vec<BackendRequest<P, R>>> {
+        log::trace!("Requested interframe timestamp={}", input_meta.timestamp);
+
+        let header = Header {
+            frame_type: FrameType::InterFrame,
+            show_frame: true,
+            error_resilient_mode: true,
+            width: self.config.resolution.width,
+            height: self.config.resolution.height,
+            render_and_frame_size_different: false,
+            render_width: self.config.resolution.width,
+            render_height: self.config.resolution.height,
+            intra_only: false,
+            refresh_frame_flags: 0x01,
+            ref_frame_idx: [0; 3],
+
+            ..Default::default()
+        };
+
+        let ref_frame = self.references.pop_front().unwrap();
```

Comment:
I don't see a hard guarantee that the `unwrap` won't crash us - it would be nice to remove it at some point.

Reply:
Oh, I should add a `// SAFETY: ` comment. It is in the `next_request()`. If `references` is empty, this function would not be called.